File: bench/py/95_recursion_with_builtin.py
# a recursion whose body calls a builtin (max) at every level (depth ~500)
import sys
sys.setrecursionlimit(10000)
scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1

def climb(n, h):
    if n <= 0:
        return h
    return climb(n - 1, max(h, (n * 37) % 1009) - 1)

K = 2000 * scale
s = 0
for k in range(K):
    s = (s + climb(500 + k % 7, k % 11)) % 1000000007
print("result:", s)

File: bench/cpp/95_recursion_with_builtin.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/95_recursion_with_builtin.my: a depth-~500
 * recursion calling std::max at every level. BENCH-FAIR (class B, as
 * 10_recursion_deep): MyLang performs a real call per level, so climb is
 * noinline with the sibling-call optimisation off - g++ would otherwise
 * turn the tail call into a loop. The depth varies with k and bench_sink
 * keeps every result live. */
#include "bench.h"
#include <algorithm>

__attribute__((noinline, optimize("no-optimize-sibling-calls")))
static long climb(long n, long h)
{
    if (n <= 0)
        return h;
    return climb(n - 1, std::max(h, (n * 37) % 1009) - 1);
}

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);
    long K = 2000L * scale;
    long s = 0;
    for (long k = 0; k < K; k++) {
        s = (s + climb(500 + k % 7, k % 11)) % 1000000007;
        bench_sink(s);
    }
    printf("result: %ld\n", s);
    return 0;
}
